### ISSUE TITLE
Make Iterant's ++ take a lazy (by-name) parameter

### ISSUE THREAD

**Issue by alexandru** [monix/monix#642]
Currently the signature for `++` for `Iterant` is this:

```scala
sealed abstract class Iterant[F[_], A] {
  //...

  def ++[B >: A](rhs: Iterant[F, B])(implicit F: Applicative[F]): Iterant[F, B]
}
```

We need to change it to take a by-name parameter:

```scala
def ++[B >: A](rhs: => Iterant[F, B])(implicit F: Applicative[F]): Iterant[F, B]
```

Note that via this change we might slow down operators that depend on `++`, so we need to provide a strict alternative. Might as well provide a function on the companion object:

```scala
object Iterant {
  // ...

  def concat[F[_], A](xs: Iterant[F, A]*): Iterant[F, A] = ???
}
```

This needs to be implemented in the most efficient way possible, so it should not be in terms of `flatMap`, but in terms of the current `++` implementation.

Also search and replace in the code-base any usage of `++` with `concat` (I don't think there are any, but should be double checked).

**Comment by Avasil** [comment]
Just commenting that one of the Scala Spree participants is working on this issue. :)

**Comment by Avasil** [comment]
Resolved in #656
